src/libraries/System.Private.CoreLib/src/System/Math.cs

```diff
@@ -113,6 +114,52 @@ public static long BigMul(int a, int b)
             return ((long)a) * b;
         }
 
+        [CLSCompliant(false)]
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static ulong BigMul(ulong a, ulong b, out ulong low)
+        {
+            if (Bmi2.X64.IsSupported)
+            {
+                ulong tmp, high;
+                unsafe
+                {
+                    high = Bmi2.X64.MultiplyNoFlags(a, b, &tmp);
+                }
+                low = tmp;
+                return high;
+            }
+            else
+            {
+                return SoftwareFallback(a, b, out low);
+            }
+
+            static ulong SoftwareFallback(ulong a, ulong b, out ulong low)
```

Comment:
Did you grab the algorithm from anywhere in particular? It would be useful to document from where, if so, and to explain the logic here.
It always takes me a minute to remember that this is basically just the FOIL method in code...

That is, the basic premise is if you take `a` as being `a1:a0` and `b` as being `b1:b0` then the algorithm below is basically: `((a1 * 2^32) + a0) + ((b1 * 2^32) + b0)`, but optimized a bit

Reply:
High qword of multiplication was adopted from 32-bit version with unsigned operand for RISC architecture from _Hacker's Delight_ book by _Henry S. Warren, Jr._ Low qword was added by me.